### ISSUE TITLE
Continue comment when enter is pressed

### ISSUE THREAD

**Issue by jonhoo** [microsoft/vscode#26694]
- VSCode Version: 1.12.1
- OS Version: Arch Linux

Steps to Reproduce:

1. Open editor in a language that supports comments (say `foo.c`)
2. Type a line that starts with `//`
3. Press `enter`
4. Next line *should* automatically be started with `//` (or at least this should be possible to configure). Instead, the next line is empty.

**Comment by PRADEEPKUMARRAJAMANICKAM** [comment]
+1 this would be nice

**Comment by rebornix** [comment]
This should be implemented in a C extension by customizing enter rules. We may want to update our internal C extension.

**Comment by noamyogev84** [comment]
Hi, can I jump in? :) new to this repo

**Comment by daltonamitchell** [comment]
@rebornix If you can point me in the right direction for where to make these changes I'm happy to submit a PR for this. I like to start new files with comments laying out the code I'm _going_ to write, so I'd get a lot of use from this.

**Comment by tobia** [comment]
Is there a way to enable this with a keybinding, for all languages?

Say, if the cursor is inside a line comment and you press enter, perform 1. enter; 2. line comment.

**Comment by ctf0** [comment]
for now u can try https://marketplace.visualstudio.com/items?itemName=ctf0.auto-comment-next-line

**Comment by peey** [comment]
There should be a configurable editor option to do this, if it's not default.

Very annoying to have to start comment on every line.

**Comment by Typz** [comment]
This feature is provided by https://marketplace.visualstudio.com/items?itemName=kevinkyang.auto-comment-blocks

**Comment by kylechadha** [comment]
Thanks @Typz. Just tried it however it doesn't work with javascript when `auto-comment-blocks.singleLineBlockOnEnter` is set to true (which is the behavior being requested here), and it doesn't look to be supported anymore (last commit was 3 years ago).

This functionality is common in other editors so it'd be great to see it added to vscode as well.

**Comment by katerlouis** [comment]
> This feature is provided by https://marketplace.visualstudio.com/items?itemName=kevinkyang.auto-comment-blocks, although disabled by default in the extension settings.

Correct me if I'm wrong, but the extension seems to only deal with comment blocks (`/** */`) and not with "normal comments"

**Comment by valeriangalliat** [comment]
The above extensions weren't working for me so I made a new one to add inline comments to `onEnterRules` of existing languages: https://marketplace.visualstudio.com/items?itemName=val.vscode-continue-inline-comments

**Comment by abettermap** [comment]
@valeriangalliat very nice! I have been wanting something like that for years. I tested it briefly with a TypeScript React file and it worked great.

I have enjoyed another great extension, [Rewrap](https://marketplace.visualstudio.com/items?itemName=stkb.rewrap), but it does not always handle new comment lines the way I want it to when using TSDoc, for example.

Both extensions have use cases though, and therefore a place in my toolbelt, so thanks for creating it 🙌

**Comment by SaltedBlowfish** [comment]
> The above extensions weren't working for me so I made a new one to add inline comments to `onEnterRules` of existing languages: https://marketplace.visualstudio.com/items?itemName=val.vscode-continue-inline-comments

Seems to work fine for most languages, but `go` doesn't work. Just FYI.

**Comment by brokenthorn** [comment]
To get this as a built-in editor setting, or language setting, a common LSP-supported approach is probably needed. The LSP server should understand the concept of a `block comment` to be able to extend that block on new line if inserted before the end of the block and after the start of the block.

Perhaps the reason this has not yet been made available as an option is because it has been overlooked, or it could very well be because the LSPs implementations don't agree on a common API that would be able to provide this.

I've never done LSP work, but from my understanding of them, it seems like this would be an LSP-centred feature.

I took would like this feature but don't know if I can find the time to learn LSP stuff (don't have too much free time right now).

**Comment by starball5** [comment]
Possible dup of #180551, which is currently in the October 2023 Milestone.

**Comment by cyanophage** [comment]
It seems that this has been implemented now and is on by default. How do I turn it off?

**Comment by katerlouis** [comment]
> It seems that this has been implemented now and is on by default. How do I turn it off?

I'll tell you, when you tell me how I can switch it on 8) – Seriously though, are we talking about the same here? 

```js
// when I press enter here, the next line is no comment :'(
const person = 'Homer Simpson';
```

**Comment by tim-hilt** [comment]
Same for me. I'm quite confused tbh :D @cyanophage are you sure that that's not a feature of a particular language-extension you're using?

**Comment by tomasklaen** [comment]
Not just continue comment, but when I press "Join lines" shortcut, it should remove the comment prefix from the joined line.

I remember having both of these in SublimeText and I still miss it even after yeas of vscode usage :(

**Comment by yuezk** [comment]
> Not just continue comment, but when I press "Join lines" shortcut, it should remove the comment prefix from the joined line.
> 
> I remember having both of these in SublimeText and I still miss it even after yeas of vscode usage :(

You got me there! I have this feature in Vim with a comment plugin, and it's super convenient. Every time I break a long comment line or join comment lines in VS Code, I think about this feature. I assumed there might be a configuration in VS Code, and finally stumbled upon this discussion.

Hope VS Code will support this feature soon.

**Comment by starball5** [comment]
Related(?) on Stack Overflow: [How to configure VSCode to auto insert comment chars on multi line comments](https://stackoverflow.com/q/43542259/11107541)

**Comment by aiday-mar** [comment]
Hi everyone, I have the following PR which should fix this issue for the following languages that use the double-slash line comment:

cpp
csharp
go
groovy
java
json
less
objective-c
php
rust
swift

TypeScript and JavaScript already have this fix. This will add a line comment on the next line when pressing Enter only if the cursor is NOT followed only by whitespaces. This was decided as per a request from users after the fix came out for TypeScript/JavaScript. In other words, in the following case where | represents the cursor position, pressing enter will not add a line comment on the next line:

`// some comment |`

But in the following case it will

`// some | comment`

I have not added this change for C because we do not have a language configuration file for C and do not support auto-indentation natively for C. If there are other languages you'd like me to cover, with potential, other types of line comment syntaxes, please file a new issue.

**Comment by sbrocket** [comment]
> This will add a line comment on the next line when pressing Enter only if the cursor is NOT followed only by whitespaces. This was decided as per a request from users after the fix came out for TypeScript/JavaScript.

Can you clarify what the purpose of this is and why it makes sense to apply to the general feature? This seems like a non-obvious special case that will confuse people, but maybe I’m missing something.

**Comment by sbrocket** [comment]
> this is as a result of the following issue https://github.com/microsoft/vscode/issues/233186 as well as comments in https://github.com/microsoft/vscode/issues/180551 and generally feedback I have received in standup about not wanting to add a line comment on the next line if the cursor is at the end of the line followed only by whitespaces

I think this tells me why it was implemented this way (user feedback) but it’s still not obvious to me why people gave  this feedback. What are the motivating use cases that make adding this exception better than not doing so?

> The thought is that generally when one wants to write a multi-line comment, they'd probably use the multi-line format /* */.

This part feels a lot more like user preference than a safe assumption to make at the editor level. It’s not obvious to me that this is true generally for me at all.

I would guess instead that it probably depends on whatever the style happens to be in an existing codebase, and so maybe even something I may want to configure at a workspace level? It’s honestly difficult for me to be sure of without sitting down and listing out use cases for both behaviors, but at minimum, if that assumption doesn’t hold true for any user then this behavior will seem non-obvious (aka “random” in confused user speak) and so it seems like it will be really annoying to those users.

> If however the cursor is within a line comment followed by non-whitespace characters, then we can assume the user wants to add a line comment on the next line.

Yes this part does seem like a safe assumption.

**Comment by aiday-mar** [comment]
Yes essentially the feedback I got from the users I talked to was that they don't want a line comment to be inserted on the next line when pressing Enter from the end of a line comment. I think these users use the line comment to add single line comments (as generally the line comment was designed moreso for this purpose). For those users it can be disturbing to have to remove the double slash on Enter every time. I am open to discussion however and if a lot of people in this thread would like to add this functionality we can discuss it.
